src/util/bson.hpp

```diff
@@ -0,0 +1,1218 @@
+        float,
+        double,
+        time_t,
+        const char*,
+        BinaryData,
+        Timestamp,
+        Decimal128,
+        ObjectId,
+        RegularExpression,
+        MinKey,
+        MaxKey,
+        IndexedMap<Bson>,
+        std::vector<Bson>
+    >;
+    using base::base;
+    using base::operator=;
+};
+
+using BsonDocument = IndexedMap<Bson>;
+using BsonArray = std::vector<Bson>;
+
+
+template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
+template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
+
+/**
+ ======= JSON Serialization and Streams =======
+ */
+namespace {
+static void to_json(const Bson& bson, std::stringstream& ss)
+{
+    std::visit(overloaded {
+        [&ss](auto) {
+            ss << "null";
+        },
+        [&ss](bson_null_t) {
+            ss << "null";
+        },
+        [&ss](int32_t bson) {
+            ss << "{" << "\"$numberInt\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&ss](double bson) {
+            ss << "{" << "\"$numberDouble\"" << ":" << '"';
+            if (isnan(bson)) {
+                ss << "NaN";
+            } else if (bson == std::numeric_limits<double>::infinity()) {
+                ss << "Infinity";
+            } else if (bson == (-1 * std::numeric_limits<double>::infinity())) {
+                ss << "-Infinity";
+            } else {
+                ss << bson;
+            }
+            ss << '"' << "}";
+        },
+        [&ss](int64_t bson) {
+            ss << "{" << "\"$numberLong\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&ss](Decimal128 bson) {
+            ss << "{" << "\"$numberDecimal\"" << ":" << '"';
+            if (bson.is_nan()) {
+                ss << "NaN";
+            } else if (bson == Decimal128("Infinity")) {
+                ss << "Infinity";
+            } else if (bson == Decimal128("-Infinity")) {
+                ss << "-Infinity";
+            } else {
+                ss << bson;
+            }
+            ss << '"' << "}";
+        },
+        [&ss](ObjectId bson) {
+            ss << "{" << "\"$oid\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&ss](BsonDocument bson) {
+            ss << "{";
+            for (auto const& pair : bson)
+            {
+                ss << '"' << pair.first << "\":";
+                to_json(pair.second, ss);
+                ss << ",";
+            }
+            if (bson.size())
+                ss.seekp(-1, std::ios_base::end);
+            ss << "}";
+        },
+        [&ss](BsonArray bson) {
+            ss << "[";
+            for (auto const& b : bson)
+            {
+                to_json(b, ss);
+                ss << ",";
+            }
+            if (bson.size())
+                ss.seekp(-1, std::ios_base::end);
+            ss << "]";
+        },
+        [&ss](BinaryData bson) {
+            ss << "{\"$binary\":{\"base64\":\"" << (bson.is_null() ? "" : bson.data()) << "\",\"subType\":\"00\"}}";
```

Comment:
This doesn't seem to actually be printing a base64 encoded version of the data?
I think this should work:
```
        util::StringBuffer encode_buffer;
        encode_buffer.resize(util::base64_encoded_size(bson.size()));
        util::base64_encode(bson.data(), bson.size(), encode_buffer.data(), encode_buffer.size());
        ss << encode_buffer.str();
```